pkg/csi/node.go

```diff
@@ -2,41 +2,192 @@ package csi
+		zap.String("bundle", bundle))
+	return &csi.NodeStageVolumeResponse{}, nil
+}
+
+func (n *nodeServer) prepBundle(repo string, bundle string, volumeId string) error {
+	// Check if the bundle has been downloaded, if not download it.
+	_, ok := n.bundleMap[getDownloadedBundleKey(repo, bundle)]
+	if !ok {
+		path, err := getPathToLocalFS(n.localFS, bundle, repo)
+		if err != nil {
+			n.l.Error("failed to create mount path", zap.Error(err))
+			return err
+		}
+		localFS := localfs.New(afero.NewBasePathFs(afero.NewOsFs(), path))
+		bd := core.NewBDescriptor()
+		b := core.New(bd,
+			core.Repo(repo),
+			core.BundleID(bundle),
+			core.BlobStore(n.blob),
+			core.ConsumableStore(localFS),
+			core.MetaStore(n.meta),
+		)
+		fs, err := core.NewReadOnlyFS(b, n.l)
+		if err != nil {
+			n.l.Error("failed to initialize bundle",
+				zap.String("repo", repo),
+				zap.String("bundle", bundle),
+				zap.Error(err))
+			return status.Error(codes.Internal, "failed to initialize repo:bundle "+repo+":"+bundle)
+		}
+		downloadedBundle := downloadedBundle{
+			repo:     repo,
+			bundleID: bundle,
+			path:     "", // TODO
+			refCount: 1,
+			fs:       fs,
+		}
+
+		n.bundleMap[volumeId] = &downloadedBundle
+		n.l.Info("volume ready to be published",
+			zap.String("volumeId", volumeId),
+			zap.String("repo", repo),
+			zap.String("bundle", bundle))
+	}
+	n.l.Info("Prep Bundle finished")
+	return nil
 }
 
 func (n *nodeServer) NodeUnstageVolume(context.Context, *csi.NodeUnstageVolumeRequest) (*csi.NodeUnstageVolumeResponse, error) {
-	return nil, nil
+	return &csi.NodeUnstageVolumeResponse{}, nil
 }
 
 func (n *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublishVolumeRequest) (*csi.NodePublishVolumeResponse, error) {
-	return nil, nil
+	n.lock.Lock()
+	defer n.lock.Unlock()
+	downloadedBundle, ok := n.bundleMap[req.VolumeId]
+	if !ok {
+		repo, ok := req.VolumeAttributes["repo"]
+		if !ok {
+			n.l.Error("repo not set for volume", zap.String("req", req.String()))
+			return nil, status.Error(codes.InvalidArgument, "repo not set for volume seen first time")
+		}
+		bundle, ok := req.VolumeAttributes["hash"]
+		if !ok {
+			n.l.Info("latest commit for main branch", zap.String("repo", repo), zap.String("req", req.String()))
+		}
+		err := n.prepBundle(repo, bundle, req.VolumeId)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		n.l.Info("mounting an existing bundle", zap.String("req", req.String()))
+	}
+	downloadedBundle, ok = n.bundleMap[req.VolumeId]
+	if !ok {
+		return nil, status.Error(codes.Internal, "fsMap missing entry: "+req.String())
+	}
+
+	err := downloadedBundle.fs.MountReadOnly(req.TargetPath)
+	if err != nil {
+		return nil, err
+	}
+
+	n.l.Info("Publish volume done",
+		zap.String("volume", req.VolumeId),
+		zap.String("id", downloadedBundle.bundleID))
+	return &csi.NodePublishVolumeResponse{}, nil
+}
+
+func getDownloadedBundleKey(repo string, bundle string) string {
+	return repo + bundle
```

Comment:
here, as in `CreateVolume`'s use of `csi.Volume.Id`, the `csi.NodePublishVolumeRequest.VolumeId` corresponds to a bundle, except there's no `-` delimiter between the repo and bundle.. is it that the k8s and CSI apis remove the delimiter, or is there a mismatch between formatting the volume ids?